llvm/lib/Transforms/Scalar/JumpThreading.cpp

```diff
@@ -1409,6 +1409,13 @@ bool JumpThreadingPass::simplifyPartiallyRedundantLoad(LoadInst *LoadI) {
     if (AATags)
       NewVal->setAAMetadata(AATags);
 
+    if (auto *MD = LoadI->getMetadata(LLVMContext::MD_invariant_load))
+      NewVal->setMetadata(LLVMContext::MD_invariant_load, MD);
+    if (auto *InvGroupMD = LoadI->getMetadata(LLVMContext::MD_invariant_group))
+      NewVal->setMetadata(LLVMContext::MD_invariant_group, InvGroupMD);
+    if (auto *RangeMD = LoadI->getMetadata(LLVMContext::MD_range))
+      NewVal->setMetadata(LLVMContext::MD_range, RangeMD);
```

Comment:
Can we copy all metadata and then call dropUBImplyingAttrsAndMetadata()? We could only call it for the !isGuaranteedToTransferExecutionToSuccessor case.

I don't think preserving invariant_load is valid for the non-guaranteed-to-transfer case, at least with the currently specified semantics.

Reply:
Copied from https://github.com/llvm/llvm-project/blob/c41ef751f85ee42f40380d064f44cfd347fb6215/llvm/lib/Transforms/Scalar/GVN.cpp#L1519-L1532.

I agree we may need a better helper function.

Reply:
>  We could only call it for the !isGuaranteedToTransferExecutionToSuccessor case.
I don't think preserving invariant_load is valid for the non-guaranteed-to-transfer case, at least with the currently specified semantics.

It is guaranteed by previous checks:
https://github.com/llvm/llvm-project/blob/35f4cdbf59fca82b97869cce7e9e5d5009144938/llvm/lib/Transforms/Scalar/JumpThreading.cpp#L1357-L1369
https://github.com/llvm/llvm-project/blob/35f4cdbf59fca82b97869cce7e9e5d5009144938/llvm/lib/Transforms/Scalar/GVN.cpp#L1727-L1730

Reply:
There are two cases here. If the instructions we're hoisting over are willreturn, then we can preserve all metadata. The use of copyMetadataForLoad() is not necessary, as we're not changing the load type. We can directly copy all metadata.

If the instructions we're hoisting over are not willreturn, but the load is speculatable, we have to be more careful and typically cannot preserve UB-implying metadata. Using copyMetadataForLoad() is incorrect in this case, as it will preserve `!noundef` for example.

Reply:
Oops, I don't know `isSafeToSpeculativelyExecute` returns true even if the load has UB-implying metadata...